Sources/_StringProcessing/ConsumerInterface.swift

```diff
@@ -51,24 +51,34 @@ extension DSLTree.Node {
 }
 
 extension DSLTree.Atom {
-  // TODO: If ByteCodeGen switches first, then this is
-  // unnecessary...
+  // TODO: If ByteCodeGen switches first, then this is unnecessary for
+  // top-level nodes, but it's also invoked for `.atom` members of a custom CC
   func generateConsumer(
     _ opts: MatchingOptions
   ) throws -> MEProgram<String>.ConsumeFunction? {
+    let isCaseInsensitive = opts.isCaseInsensitive
+    
     switch self {
-
     case let .char(c):
       // TODO: Match level?
       return { input, bounds in
         let low = bounds.lowerBound
-        guard input[low] == c else {
-          return nil
+        if isCaseInsensitive && c.isCased {
+          return input[low].lowercased() == c.lowercased()
+            ? input.index(after: low)
+            : nil
+        } else {
+          return input[low] == c
+            ? input.index(after: low)
+            : nil
         }
-        return input.index(after: low)
       }
     case let .scalar(s):
-      return consumeScalar { $0 == s }
+      return consumeScalar {
+        isCaseInsensitive
+          ? $0.properties.lowercaseMapping == s.properties.lowercaseMapping
+          : $0 == s
```

Comment:
I still feel like there's quite a bit of duplicated code throughout. Would a helper function, similar to how `consumeScalar` is a helper function for making predicate-based scalar consumers, alleviate this some? Maybe something that is parameterized over case sensitivity?